google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/BitReversedSequenceStyleGenerator.java

```diff
@@ -168,39 +249,46 @@ protected DatabaseStructure buildTableStructure(
   public Serializable generate(SharedSessionContractImplementor session, Object object)
       throws HibernateException {
     Serializable id;
-    int attempts = 0;
-    // Loop to retry the transaction that updates the table-backed sequence if it fails due to a
-    // concurrent modification error. This can happen if multiple entities are using the same
-    // sequence for generated primary keys.
+    // Loop to skip excluded ranges.
     while (true) {
-      try {
-        id = generateBaseValue(session, object);
-        break;
-      } catch (GenericJDBCException exception) {
-        JdbcAbortedDueToConcurrentModificationException aborted;
-        if (exception.getSQLException()
-            instanceof JdbcAbortedDueToConcurrentModificationException) {
-          aborted = (JdbcAbortedDueToConcurrentModificationException) exception.getSQLException();
-        } else {
-          throw exception;
-        }
-        attempts++;
-        if (attempts == MAX_ATTEMPTS) {
-          throw exception;
-        }
+      int attempts = 0;
+      // Loop to retry the transaction that updates the table-backed sequence if it fails due to a
+      // concurrent modification error. This can happen if multiple entities are using the same
+      // sequence for generated primary keys.
+      while (true) {
         try {
-          sleep(aborted.getCause().getRetryDelayInMillis());
-        } catch (InterruptedException interruptedException) {
-          Thread.currentThread().interrupt();
-          throw new IdentifierGenerationException("Interrupted while trying to generate a new ID",
-              interruptedException);
+          id = generateBaseValue(session, object);
+          break;
+        } catch (GenericJDBCException exception) {
+          JdbcAbortedDueToConcurrentModificationException aborted;
+          if (exception.getSQLException()
+              instanceof JdbcAbortedDueToConcurrentModificationException) {
+            aborted = (JdbcAbortedDueToConcurrentModificationException) exception.getSQLException();
+          } else {
+            throw exception;
+          }
+          attempts++;
+          if (attempts == MAX_ATTEMPTS) {
+            throw exception;
+          }
+          try {
+            sleep(aborted.getCause().getRetryDelayInMillis());
+          } catch (InterruptedException interruptedException) {
+            Thread.currentThread().interrupt();
+            throw new IdentifierGenerationException(
+                "Interrupted while trying to generate a new ID", interruptedException);
+          }
         }
       }
+      if (id instanceof Long) {
+        Long reversed = Long.reverse((Long) id);
+        if (excludeRanges.stream().noneMatch(range -> range.contains(reversed))) {
```

Comment:
Would be good to add info logs when we couldn't generate ID due to excluded ranges? Probably in the `else` block?